azure-iot-device/azure/iot/device/common/handle_exceptions.py

```diff
@@ -24,7 +25,8 @@ def handle_background_exception(e):
 
     # @FUTURE: We should add a mechanism which allows applications to receive these
     # exceptions so they can respond accordingly
-    logger.error(msg="Exception caught in background thread.  Unable to handle.", exc_info=e)
+    logger.error(msg="Exception caught in background thread.  Unable to handle.")
+    logger.error(traceback.format_exception_only(type(e), e))
```

Comment:
>logger.error(traceback.format_exception_only(type(e), e)) [](start = 4, length = 57)

Very subtle but important difference here.  Horton checks for memory leaks after every test, but the old code was causing something that looked like a memory leak (but wasn't actually a leak per-se).

When you use exc_info=e or exc_info=True, it causes the Exception itself, including the traceback, to be saved by pytest until the test is done.  The problem with tracebacks is that they hold onto the actual objects in the stack, so everything in the stack "leaks" until the test is done.

This "false leak report" by itself could be fixed by better leak detection, but the other side effect (that we don't see yet) is that object lifetime is far longer than it needs to be (but only when running under pytest).  So, if we had a long-run-time test that ran under pytest, any objects associated with any exceptions (even when handled correctly) would not be collected until the end of the run.